metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py

```diff
@@ -0,0 +1,911 @@
+        """
+        Get user for the given PowerBi entity
+        """
+        users: List[PowerBiAPI.User] = []
+        if self.__config.extract_ownership is False:
+            LOGGER.info(
+                "ExtractOwnership capabilities is disabled from configuration and hence returning empty users list"
+            )
+            return users
+
+        user_list_endpoint: str = PowerBiAPI.API_ENDPOINTS[Constant.ENTITY_USER_LIST]
+        # Replace place holders
+        user_list_endpoint = user_list_endpoint.format(
+            POWERBI_ADMIN_BASE_URL=PowerBiAPI.ADMIN_BASE_URL,
+            ENTITY=entity,
+            ENTITY_ID=_id,
+        )
+        # Hit PowerBi
+        LOGGER.info(f"Request to URL={user_list_endpoint}")
+        response = requests.get(
+            user_list_endpoint,
+            headers={Constant.Authorization: self.get_access_token()},
+        )
+
+        # Check if we got response from PowerBi
+        if response.status_code != 200:
+            LOGGER.warning(
+                "Failed to fetch user list from power-bi. http_status=%s. message=%s",
+                response.status_code,
+                response.text,
+            )
+
+            LOGGER.info(f"{Constant.WorkspaceId}={workspace_id}")
+            LOGGER.info(f"{Constant.ENTITY}={entity}")
+            LOGGER.info(f"{Constant.ID}={_id}")
+            raise ConnectionError("Failed to fetch the user list from the power-bi")
+
+        users_dict: List[Any] = response.json()[Constant.VALUE]
+
+        # Iterate through response and create a list of PowerBiAPI.Dashboard
+        users = [
+            PowerBiAPI.User(
+                id=instance.get("identifier"),
+                displayName=instance.get("displayName"),
+                emailAddress=instance.get("emailAddress"),
+                graphId=instance.get("graphId"),
+                principalType=instance.get("principalType"),
+            )
+            for instance in users_dict
+        ]
+
+        return users
+
+    def __get_report(
+        self, workspace_id: str, report_id: str
+    ) -> Optional["PowerBiAPI.Report"]:
+        """
+        Fetch the report from PowerBi for the given report identifier
+        """
+        if workspace_id is None or report_id is None:
+            LOGGER.info("Input values are None")
+            LOGGER.info(f"{Constant.WorkspaceId}={workspace_id}")
+            LOGGER.info(f"{Constant.ReportId}={report_id}")
+            return None
+
+        report_get_endpoint: str = PowerBiAPI.API_ENDPOINTS[Constant.REPORT_GET]
+        # Replace place holders
+        report_get_endpoint = report_get_endpoint.format(
+            POWERBI_BASE_URL=PowerBiAPI.BASE_URL,
+            WORKSPACE_ID=workspace_id,
+            REPORT_ID=report_id,
+        )
+        # Hit PowerBi
+        LOGGER.info(f"Request to report URL={report_get_endpoint}")
+        response = requests.get(
```

Comment:
minor: should we add retry logic? 

is there any scenario in which this method throws an exception? perhaps we should try: except on these external calls

Reply:
It only fails if the server is not reachable, in such cases user can re-run the ingestion.